### PR TITLE
Moved validation logic to script parameters; moved all download handi…

### PR DESCRIPTION
**Pull request by kroteau**
Revised logic for:
Parameter validation, download handling, file handling, process handling, logging
Added verbose output
p.s.: some of the code is untested as windows guest mode does not support win32_product wmi call and integration runtime is not able to install due to power management policy

### CODE REVIEW

File: install-gateway.ps1 [hunk 1]
```diff
@@ -1,79 +1,144 @@
...
+  {
+    $p1 = Start-Process -FilePath $executable -ArgumentList @("-EnableRemoteAccess", "8060") -Wait -PassThru
+    $p2 = Start-Process -FilePath $executable -ArgumentList @("-RegisterNewNode", "$key", "$env:COMPUTERNAME") -Wait -PassThru
+  }
+
+  switch($p1.exitCode + $p2.exitCode)
+  {
+    0
+      {
+        Write-Host "Succeed to register the gateway"
```
Comment (favoretti) [inline install-gateway.ps1]:
```suggestion
        Write-Host "Registration with the gateway was successul"
```

File: install-gateway.ps1 [hunk 2]
```diff
@@ -1,79 +1,144 @@
...
+      }
+    $null
+      {
+        Write-Host "Register gateway impossible, executable not found"
```
Comment (favoretti) [inline install-gateway.ps1]:
```suggestion
        Write-Host "Unable to register with the gateway, executable not found"
```

File: install-gateway.ps1 [hunk 3]
```diff
@@ -1,79 +1,144 @@
...
-  $destination = "$workd\vcredist_x64.exe"
-    $client = New-Object System.Net.WebClient
-    $client.DownloadFile("https://download.microsoft.com/download/3/2/2/3224B87F-CFA0-4E70-BDA3-3DE650EFEBA5/vcredist_x64.exe", $destination)
-    Start-Process -FilePath "$workd\vcredist_x64.exe" -ArgumentList "/Q" -Wait
+  $filename = Download-File -uri $vcrsource -msdl_filter $vcrsource_filter
+
+  Write-Verbose "Installing VcRedist from $filename"
+  $process = Start-Process -FilePath "$filename" -ArgumentList "/Q /log $workd\vcredist.log" -Wait -PassThru
+  
+  Start-Sleep 2
```
Comment (favoretti) [inline install-gateway.ps1]: I see that you made a neater process waiting in JRE install, can we do the same here please? Also some max timeout might be useful, so that we don't wait forever and fail after a certain time.

File: install-gateway.ps1 [hunk 4]
```diff
@@ -1,79 +1,144 @@
...
 # uninstall any existing gateway
   UnInstall-Gateway
 
-    Write-Host "Start Gateway installation"
+  Write-Host 'Start Gateway installation'
+  
+  Write-Verbose "Installing Gateway from $filename"
+  
+  $process = Start-Process -FilePath "msiexec.exe" -ArgumentList "/i $filename /quiet /passive /L*v $workd\ir.log" -Wait -PassThru
+  Start-Sleep -Seconds 30
```
Comment (favoretti) [inline install-gateway.ps1]: Same here, van we wrap the waiter so it waits until the process is gone? Also some max timeout might be useful, so that we don't wait forever and fail after a certain time.

**Review by favoretti — CHANGES_REQUESTED** [review]
Thanks @kroteau!

A few remarks, but otherwise this is looking good!
Did you happen to check backwards-compatibility?

File: install-gateway.ps1 (hunk 3 above)
Comment (kroteau) [reply install-gateway.ps1]: VCredist uses a different kind of installer that does not need a workaround in place. Basically when you say "-Wait" it will wait untill the installer is finished or failed and the $process variable will receive the execution information upon the installer either succeed or fail. On the other hand JRE installer drops connection to the console and _kinda_ goes its own way, so we need a separate process to tell if it is completed (or not). The timeout is actually a great idea, I'll work on that one!

File: install-gateway.ps1 (hunk 1 above)
Comment (kroteau) [reply install-gateway.ps1]: I thought on removing all the write-host blocks altogether, as now there is a special function to handle operation results with fair amount of information. I should have posted a sample output for a reference.

File: install-gateway.ps1 (hunk 4 above)
Comment (kroteau) [reply install-gateway.ps1]: Same here, MSI installers handle "-Wait" flag pretty good, the sleep timer is mostly for keeping the process extra safe, like edge cases when COM+ or registry objects could not complete their transactions to proceed with the gateway registration being 101% sure that it will work as expected

**Comment by kroteau** [comment]
> Thanks @kroteau! 
> A few remarks, but otherwise this is looking good! Did you happen to check backwards-compatibility?

As you can see there are several force-push actions after I've opened the PR. So yeah, it should handle both the old way to do things and the new way perfectly. Backwards compatibility caught my eye when I've started to work on the upcoming TF module version. On my first commit we were a little screwed-up if the first argument (which is the mandatory one in the previous version) goes away. If you look on the file header - the ParameterSet* stuff is there to shift expectations on the fly. If the first argument is a file path - we're dropping to the _legacy_ way and installing from the file and if there is only one parameter which is a sting - we're good, it is a key "go grab the latest version from the downloads page" routine.
In conclusion - I've made sure that we should be good on that one :)

TODO: leave a sufficient PR description next time

**Comment by kroteau** [comment]
(I should have done this a week ago) this is the new way to call the script (there is an issue that I cannot test IR install fully as windows guest mode does not expose win32_product for some reason, but still)
```
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 authkey123
                                                                                                                                                                                                                                                                                                                                                Directory: C:\                                                                                                                                                                                                                                                                                                                          
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----         5/31/2023   2:56 AM                temp
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
ExitCode=0; StartTime=05/31/2023 02:56:25; ExitTime=05/31/2023 02:56:32
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11
Waiting for jreinstall.exe to finish...
ExitCode=0; StartTime=05/31/2023 02:56:46; ExitTime=05/31/2023 02:57:29
FileName=C:\temp\IntegrationRuntime_5.29.8528.1.msi; FileVersion=; ProductVersion=
Microsoft Integration Runtime not found.
Start Gateway installation
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=05/31/2023 02:59:51; ExitTime=05/31/2023 03:03:03
Start to register gateway with key: authkey123
Register gateway impossible, executable not found
```
Notice the creation of c:\temp is also present in the output.

This is the "legacy mode" usage. The only difference in "non-verbose" mode is in the script arguments: 
```
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 C:\temp\IntegrationRuntime_5.29.8528.1.msi authkey123
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
ExitCode=0; StartTime=05/31/2023 03:13:45; ExitTime=05/31/2023 03:13:51
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11
Waiting for jreinstall.exe to finish...
ExitCode=0; StartTime=05/31/2023 03:14:08; ExitTime=05/31/2023 03:14:58
FileName=C:\temp\IntegrationRuntime_5.29.8528.1.msi; FileVersion=; ProductVersion=
Microsoft Integration Runtime not found.
Start Gateway installation
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=05/31/2023 03:17:24; ExitTime=05/31/2023 03:21:22
Start to register gateway with key: authkey123
Register gateway impossible, executable not found
```

This is an example of verbose output.
It is a bit too "talky", but gives perfect perspective on what's actually happening in the moment including working mode, downloads, urls, states, and couple of 'ls'
```
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 authkey123 -Verbose
VERBOSE: Parameters:
Key              Value
---              -----
Verbose          True
authKey          authkey123
path
workd            c:\temp
parameterSetName LatestVersion


VERBOSE: Searching for "https://download.microsoft.com*vcredist_x64.exe" on download page https://www.microsoft.com/en-us/download/confirmation.aspx?id=26999
VERBOSE: GET with 0-byte payload
VERBOSE: received 154146-byte response of content type text/html
VERBOSE: Downloading https://download.microsoft.com/download/1/6/5/165255E7-1014-4D0A-B094-B6A430A6BFFC/vcredist_x64.exe into c:\temp\vcredist_x64.exe
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
VERBOSE: Installing VcRedist from c:\temp\vcredist_x64.exe
ExitCode=0; StartTime=05/31/2023 03:04:30; ExitTime=05/31/2023 03:04:37
VERBOSE: Downloading https://javadl.oracle.com/webapps/download/AutoDL?BundleId=245479_4d5417147a92418ea8b615e228bb6935 into c:\temp\jreInstall.exe
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11
VERBOSE: Install JRE from c:\temp\jreInstall.exe
Waiting for jreinstall.exe to finish...
ExitCode=0; StartTime=05/31/2023 03:04:52; ExitTime=05/31/2023 03:05:41
VERBOSE: Searching for "https://download.microsoft.com*msi" on download page https://www.microsoft.com/en-us/download/confirmation.aspx?id=39717
VERBOSE: GET with 0-byte payload
VERBOSE: received 155258-byte response of content type text/html
VERBOSE: Downloading https://download.microsoft.com/download/E/4/7/E4771905-1079-445B-8BF9-8A1A075D8A10/IntegrationRuntime_5.29.8528.1.msi into
c:\temp\IntegrationRuntime_5.29.8528.1.msi
FileName=C:\temp\IntegrationRuntime_5.29.8528.1.msi; FileVersion=; ProductVersion=
Microsoft Integration Runtime not found.
Start Gateway installation
VERBOSE: Installing Gateway from c:\temp\IntegrationRuntime_5.29.8528.1.msi
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=05/31/2023 03:08:23; ExitTime=05/31/2023 03:11:31
Start to register gateway with key: authkey123
Register gateway impossible, executable not found
VERBOSE:

    Directory: C:\temp


Mode                 LastWriteTime         Length Name

----                 -------------         ------ ----

-a----         5/31/2023   3:08 AM      994455552 IntegrationRuntime_5.29.8528.1.msi

-a----         5/31/2023   3:11 AM       12520908 ir.log

-a----         5/31/2023   3:04 AM         340784 vcredist.log-Microsoft Visual C++ 2010  x64 Redistributable - 10.0.40219-MSP0.txt

-a----         5/31/2023   3:04 AM         636474 vcredist.log-MSI_vc_red.msi.txt

-a----         5/31/2023   3:04 AM         163724 vcredist.log.html

-a----         5/31/2023   3:04 AM       10277328 vcredist_x64.exe
```
p.s.: I've also spotted an issue with the legacy mode, will return to you with the fix around it (hopefully) shortly

**Comment by kroteau** [comment]
Made some changes to the script:
- now compatibility is the default mode (and it actually works, I've tested it twice)
- "get me the latest IR" mode now requires a named parameter to be specified. That's unfortunate, but parameter validation fights with positional parameters and loses.
- figured out the proper way to generate "start report" in verbose mode, now it includes script parameters dynamically
- slightly changed (and made worse?) switching logic between legacy and compat modes, the only other option is to drop some 'if's which is even worse

Some logs:
```
# compatibility
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 C:\temp\IntegrationRuntime_5.29.8528.1.msi key123
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
ExitCode=0; StartTime=06/02/2023 03:01:35; ExitTime=06/02/2023 03:01:41                                
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11                   
Waiting for jreinstall.exe to finish...                                                                
ExitCode=0; StartTime=06/02/2023 03:01:53; ExitTime=06/02/2023 03:02:41                                
Microsoft Integration Runtime not found.
Start Gateway installation
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=06/02/2023 03:02:42; ExitTime=06/02/2023 03:07:16
Start to register gateway with key: key123
Register gateway impossible, executable not found
```
```
# latest
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 -authKey key123
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
ExitCode=0; StartTime=06/02/2023 03:16:32; ExitTime=06/02/2023 03:16:39
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11
Waiting for jreinstall.exe to finish...
ExitCode=0; StartTime=06/02/2023 03:16:52; ExitTime=06/02/2023 03:17:41
FileName=C:\temp\IntegrationRuntime_5.29.8528.1.msi; FileVersion=; ProductVersion=
Microsoft Integration Runtime not found.
Start Gateway installation
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=06/02/2023 03:19:35; ExitTime=06/02/2023 03:22:50
Start to register gateway with key: key123
Register gateway impossible, executable not found
```
the verbose logs look like this now:
```
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 C:\temp\IntegrationRuntime_5.29.8528.1.msi key123 -Verbose
VERBOSE:
ParameterSetName:  Compatibility
ParameterCount:    3
PsBoundParameters: [Verbose, True] [path, C:\temp\IntegrationRuntime_5.29.8528.1.msi] [authKey1, key123]
Script parameters (and their defaults):
Name               Value
----               -----
path               C:\temp\IntegrationRuntime_5.29.8528.1.msi
authKey
authKey1           key123
jresource          https://javadl.oracle.com/webapps/download/AutoDL?BundleId=245479_4d5417147a92418ea8b615e228bb6935
jreinstall_timeout 5
irsource           https://www.microsoft.com/en-us/download/confirmation.aspx?id=39717
irsource_filter    https://download.microsoft.com*msi
vcrsource          https://www.microsoft.com/en-us/download/confirmation.aspx?id=26999
vcrsource_filter   https://download.microsoft.com*vcredist_x64.exe
workd              c:\temp


VERBOSE: Searching for "https://download.microsoft.com*vcredist_x64.exe" on download page https://www.microsoft.com/en-us/download/confirmation.aspx?id=26999
VERBOSE: GET with 0-byte payload
VERBOSE: received 154021-byte response of content type text/html
VERBOSE: Downloading https://download.microsoft.com/download/1/6/5/165255E7-1014-4D0A-B094-B6A430A6BFFC/vcredist_x64.exe into c:\temp\vcredist_x64.exe
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
VERBOSE: Installing VcRedist from c:\temp\vcredist_x64.exe
ExitCode=0; StartTime=06/02/2023 03:34:35; ExitTime=06/02/2023 03:34:42
VERBOSE: Downloading https://javadl.oracle.com/webapps/download/AutoDL?BundleId=245479_4d5417147a92418ea8b615e228bb6935 into c:\temp\jreInstall.exe
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11
VERBOSE: Install JRE from c:\temp\jreInstall.exe
Waiting for jreinstall.exe to finish...
ExitCode=0; StartTime=06/02/2023 03:34:57; ExitTime=06/02/2023 03:35:46
Microsoft Integration Runtime not found.
Start Gateway installation
VERBOSE: Installing Gateway from C:\temp\IntegrationRuntime_5.29.8528.1.msi
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=06/02/2023 03:35:47; ExitTime=06/02/2023 03:39:05
Start to register gateway with key: key123
Register gateway impossible, executable not found
VERBOSE: Directory: C:\temp


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----          6/2/2023   3:19 AM      994455552 IntegrationRuntime_5.29.8528.1.msi
-a----          6/2/2023   3:39 AM       12519440 ir.log
-a----          6/2/2023   3:34 AM        2722836 vcredist.log-Microsoft Visual C++ 2010  x64
                                                  Redistributable - 10.0.40219-MSP0.txt
-a----          6/2/2023   3:34 AM        2357434 vcredist.log-MSI_vc_red.msi.txt
-a----          6/2/2023   3:34 AM         687620 vcredist.log.html
-a----          6/2/2023   3:34 AM       10277328 vcredist_x64.exe
```
and the same for the latest:
```
PS C:\Users\WDAGUtilityAccount> .\install-gateway.ps1 -authKey key123 -Verbose
VERBOSE:
ParameterSetName:  Latest
ParameterCount:    2
PsBoundParameters: [authKey, key123] [Verbose, True]                                                                                                                  Script parameters (and their defaults):                                                                                                                               
Name               Value                                                                                                                                              ----               -----                                                                                                                                              path
authKey            key123
authKey1
jresource          https://javadl.oracle.com/webapps/download/AutoDL?BundleId=245479_4d5417147a92418ea8b615e228bb6935
jreinstall_timeout 5
irsource           https://www.microsoft.com/en-us/download/confirmation.aspx?id=39717
irsource_filter    https://download.microsoft.com*msi
vcrsource          https://www.microsoft.com/en-us/download/confirmation.aspx?id=26999
vcrsource_filter   https://download.microsoft.com*vcredist_x64.exe
workd              c:\temp


VERBOSE: Searching for "https://download.microsoft.com*vcredist_x64.exe" on download page https://www.microsoft.com/en-us/download/confirmation.aspx?id=26999
VERBOSE: GET with 0-byte payload
VERBOSE: received 154021-byte response of content type text/html
VERBOSE: Downloading https://download.microsoft.com/download/1/6/5/165255E7-1014-4D0A-B094-B6A430A6BFFC/vcredist_x64.exe into c:\temp\vcredist_x64.exe
FileName=C:\temp\vcredist_x64.exe; FileVersion=10.0.40219.325; ProductVersion=10.0.40219.325
VERBOSE: Installing VcRedist from c:\temp\vcredist_x64.exe
ExitCode=0; StartTime=06/02/2023 03:42:21; ExitTime=06/02/2023 03:42:27
VERBOSE: Downloading https://javadl.oracle.com/webapps/download/AutoDL?BundleId=245479_4d5417147a92418ea8b615e228bb6935 into c:\temp\jreInstall.exe
FileName=C:\temp\jreInstall.exe; FileVersion=8.0.3110.11; ProductVersion=8.0.3110.11
VERBOSE: Install JRE from c:\temp\jreInstall.exe
Waiting for jreinstall.exe to finish...
ExitCode=0; StartTime=06/02/2023 03:42:42; ExitTime=06/02/2023 03:43:32
VERBOSE: Searching for "https://download.microsoft.com*msi" on download page https://www.microsoft.com/en-us/download/confirmation.aspx?id=39717
VERBOSE: GET with 0-byte payload
VERBOSE: received 155135-byte response of content type text/html
VERBOSE: Downloading https://download.microsoft.com/download/E/4/7/E4771905-1079-445B-8BF9-8A1A075D8A10/IntegrationRuntime_5.29.8528.1.msi into
c:\temp\IntegrationRuntime_5.29.8528.1.msi
FileName=C:\temp\IntegrationRuntime_5.29.8528.1.msi; FileVersion=; ProductVersion=
Microsoft Integration Runtime not found.
Start Gateway installation
VERBOSE: Installing Gateway from c:\temp\IntegrationRuntime_5.29.8528.1.msi
Gateway installer finished with exitCode: 1603
ExitCode=1603; StartTime=06/02/2023 03:45:25; ExitTime=06/02/2023 03:49:01
Start to register gateway with key: key123
Register gateway impossible, executable not found
VERBOSE: Directory: C:\temp


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----          6/2/2023   3:45 AM      994455552 IntegrationRuntime_5.29.8528.1.msi
-a----          6/2/2023   3:49 AM       12516812 ir.log
-a----          6/2/2023   3:42 AM        3063264 vcredist.log-Microsoft Visual C++ 2010  x64
                                                  Redistributable - 10.0.40219-MSP0.txt
-a----          6/2/2023   3:42 AM        2603440 vcredist.log-MSI_vc_red.msi.txt
-a----          6/2/2023   3:42 AM         762690 vcredist.log.html
```
Unfortunately my main testing rig is still a windows sandbox that lacks certain features. But it should be fine.

**Comment by kroteau** [comment]
Made couple of optimizations on top w/o changing any functionality

**Review by favoretti — APPROVED** [review]
Thanks @kroteau, LGTM.
